README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# iramuteqlike <a href='https://gipso.github.io/iramuteqlike/'><img src='man/figures/logo.png' align="right" height="139" /></a>

<!-- badges: start -->

[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![R-CMD-check](https://github.com/gipso/iramuteqlike/workflows/R-CMD-check/badge.svg)](https://github.com/gipso/iramuteqlike/actions)
[![Codecov test
coverage](https://codecov.io/gh/gipso/iramuteqlike/branch/main/graph/badge.svg)](https://app.codecov.io/gh/gipso/iramuteqlike?branch=main)
[![License:
MIT](https://img.shields.io/badge/license-MIT-green)](https://choosealicense.com/licenses/mit/)
[![Contributor
Covenant](https://img.shields.io/badge/Contributor%20Covenant-v2.0%20adopted-ff69b4.svg)](https://gipso.github.io/iramuteqlike/CODE_OF_CONDUCT.html)
[![Buy Me A Coffee donate
button](https://img.shields.io/badge/buy%20me%20a%20coffee-donate-yellow.svg)](https://ko-fi.com/danielvartan)
<!-- badges: end -->

## Overview

`iramuteqlike` is an R package with tools to reproduce the
[IRaMuTeQ](http://www.iramuteq.org/) software analyses in R. The aim of
`iramuteqlike` is to facilitate text analyses, in addition to
contributing to research reproducibility.

The package use different methods to compute some analyses, hence, the
results can differ from the [IRaMuTeQ](http://www.iramuteq.org/)
software.

Please note that `iramuteqlike` is not related in any way with the
[IRaMuTeQ](http://www.iramuteq.org/) developer team.

## Installation

You can install `iramuteqlike` from GitHub with:

``` r
# install.packages("remotes")
remotes::install_github("gipso/iramuteqlike")
```

## Citation

If you use `iramuteqlike` in your research, please consider citing it.
We put a lot of work to build and maintain a free and open-source R
package. You can find the `iramuteqlike` citation below.

``` r
citation("iramuteqlike")
#> 
#> To cite {iramuteqlike} in publications use:
#> 
#>   Vartanian, D., Souza, A. A., & Pedrazzoli, M. (2021). {iramuteqlike}:
#>   an R package with tools to reproduce the IRaMuTeQ software analyses
#>   in R. https://gipso.github.io/iramuteqlike/
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Unpublished{,
#>     title = {{iramuteqlike}: an R package with tools to reproduce the IRaMuTeQ software analyses in R},
#>     author = {Daniel Vartanian and Alesandra Araujo {de Souza} and Mario Pedrazzoli},
#>     year = {2021},
#>     url = {https://gipso.github.io/iramuteqlike/},
#>     note = {Lifecycle: experimental},
#>   }
```

## Contributing

`iramuteqlike` is a community project, everyone is welcome to
contribute. Take a moment to review our [Guidelines for
Contributing](https://gipso.github.io/iramuteqlike/CONTRIBUTING.html).

Please note that `iramuteqlike` is released with a [Contributor Code of
Conduct](https://gipso.github.io/iramuteqlike/CODE_OF_CONDUCT.html). By
contributing to this project, you agree to abide by its terms.

## Support `iramuteqlike`

[![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/danielvartan)

Working with science in Brazil is a daily challenge. There are few
funding opportunities available and their value is not enough to live
on. Added to this, every day Brazilian science suffers from deep cuts in
funding, which requires researchers to always look for other sources of
income.

If this package helps you in any way or you simply want to support the
author’s work, please consider donating or even creating a membership
subscription (if you can!). Your support will help with the author’s
scientific pursuit and with the package maintenance.

To make a donation click on the [Ko-fi](https://ko-fi.com/danielvartan)
button above. Please indicate the `iramuteqlike` package in your
donation message.

Thank you!


NEWS.md
<!--- https://devguide.ropensci.org/releasing.html -->
<!--- https://style.tidyverse.org/news.html -->
<!--- https://semver.org/ -->

# iramuteqlike (development version)

* Some typos in the documentation were fixed.

# iramuteqlike 0.1.0 (2021-11-08)

* First `iramuteqlike` release. 🎉

# iramuteqlike 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(checkmate)
library(friends)
library(iramuteqlike)
library(stringi)
library(testthat)

test_check("iramuteqlike")

# # For development use only (comment the code after use (Ctrl + Shift + C)).
#
# .rs.restartR()
# BiocCheck::BiocCheck()
# cffr::cff_write()
# codemetar::write_codemeta()
# covr::package_coverage()
# devtools::check()
# devtools::document()
# devtools::install()
# devtools::release()
# devtools:test()
# devtools::check_win_devel()
# goodpractice::gp()
# normalizePath(readClipboard(), "/", mustWork = FALSE)
# template <- list(package = "rotemplate")
# pkgdown::build_article("mctq", override = list(template = template))
# pkgdown::build_favicons(overwrite = TRUE)
# pkgdown::build_reference(override = list(template = template))
# pkgdown::build_site(devel = FALSE, override = list(template = template))
# cran_prep <- rhub::check_for_cran()
# spelling::spell_check_package()
# spelling::update_wordlist()
# urlchecker::url_check()
# usethis::use_coverage()
# usethis::use_github_action("check-pak")
# usethis::use_github_action("test-coverage-pak")
# usethis::use_logo("./.archive/ai/hex-logo.png")


tests/testthat/test-utils.R
test_that("placeholder() | general test", {
    expect_message(placeholder())
})


tests/testthat/test-plot_wordcloud.R
test_that("plot_wordcloud() | general test", {
    suppressMessages(suppressWarnings(
        expect_message(plot_wordcloud(
        word_frequency(head(friends::friends[[1]], 100))))
    ))
    suppressMessages(suppressWarnings(
    expect_null(plot_wordcloud(
        word_frequency(head(friends::friends[[1]], 100))))
    ))
})

test_that("plot_wordcloud() | error test", {
    data <- word_frequency(head(friends::friends[[1]], 100))

    expect_error(plot_wordcloud(1, min_freq = 1, max_words = 1,
                                random_order = TRUE, rot_per = 1,
                                colors = "#440154FF"),
                 "Assertion on 'data' failed")

    expect_error(plot_wordcloud(data, min_freq = "a", max_words = 1,
                                random_order = TRUE, rot_per = 1,
                                colors = "#440154FF"),
                 "Assertion on 'min_freq' failed")

    expect_error(plot_wordcloud(data, min_freq = 1, max_words = "a",
                                random_order = TRUE, rot_per = 1,
                                colors = "#440154FF"),
                 "Assertion on 'max_words' failed")

    expect_error(plot_wordcloud(data, min_freq = 1, max_words = 1,
                                random_order = 1, rot_per = 1,
                                colors = "#440154FF"),
                 "Assertion on 'random_order' failed")

    expect_error(plot_wordcloud(data, min_freq = 1, max_words = 1,
                                random_order = TRUE, rot_per = "a",
                                colors = "#440154FF"),
                 "Assertion on 'rot_per' failed")

    expect_error(plot_wordcloud(data, min_freq = 1, max_words = 1,
                                random_order = TRUE, rot_per = 1,
                                colors = 1),
                 "Assertion on 'colors' failed")

    expect_error(plot_wordcloud(data, min_freq = 1, max_words = 1,
                                random_order = TRUE, rot_per = 1,
                                colors = "440154FF"),
                 "Assertion on 'colors' failed")
})


R/plot_wordcloud.R
#' Plot a word cloud
#'
#' @description
#'
#' `r lifecycle::badge("experimental")`
#'
#' `plot_wordcloud()` plots a word cloud using an output of [word_frequency()].
#'
#' This function uses [wordcloud()][wordcloud::wordcloud()] to generate
#' word cloud graphics.
#'
#' @param data A [`tibble`][tibble::tibble()] object with two columns: `word`
#'   (with an unique set of words) and `freq` (with the absolute frequency of a
#'   word in `word`). Use [word_frequency()] to generate this value.
#' @param min_freq (optional) a number indicating the minimum frequency that a
#'   word can have to be plotted (default: `2`).
#' @param max_words (optional) a number indicating the maximum number of words
#'   to be plotted (default: `100`).
#' @param random_order (optional) a [`logical`][base::logical()] value
#'   indicating if the function must plot words in random order. If `FALSE`,
#'   the words will be plotted in decreasing frequency. See
#'   [wordcloud()][wordcloud::wordcloud()] to learn more (default: `FALSE`).
#' @param rot_per (optional) a number indicating the proportion words with 90
#'   degree rotation (default: `0.35`).
#' @param colors (optional) a [`character`][base::character()] object with Hex
#'   color codes indicating the color that the words must have, from least to
#'   most frequent. If `NULL`, the [viridis plasma palette][viridis::plasma()]
#'   will be used (default: `NULL`).
#' @param ... (optional) additional parameters to be passed to the
#' [wordcloud()][wordcloud::wordcloud()] function.
#'
#' @return An invisible `NULL`. This function don't aim to return values.
#'
#' @family Plot functions
#' @export
#'
#' @examples
#' if (requireNamespace("friends", quietly = TRUE)) {
#'     data <- word_frequency(head(friends::friends[[1]], 100))
#'     suppressMessages(plot_wordcloud(data))
#'
#'     ## Note that graphics rendered in the documentation may
#'     ## not represent the quality of the output.
#' }
#'
#' if (requireNamespace("stringi", quietly = TRUE)) {
#'     data <- word_frequency(stringi::stri_rand_lipsum(5))
#'     suppressMessages(plot_wordcloud(data))
#'
#'     ## Note that graphics rendered in the documentation may
#'     ## not represent the quality of the output.
#' }
plot_wordcloud <- function(data, min_freq = 2, max_words = 100,
                           random_order = FALSE, rot_per = 0.35,
                           colors = NULL, ...) {
    checkmate::assert_data_frame(data, min.rows = 1, min.cols = 2)
    checkmate::assert_names(names(data), identical.to = c("word", "freq"))
    checkmate::assert_number(min_freq, lower = 1)
    checkmate::assert_number(max_words, lower = 1)
    checkmate::assert_flag(random_order)
    checkmate::assert_number(rot_per)
    checkmate::assert_character(colors, min.len = 1, pattern = "^#",
                                null.ok = TRUE)

    if (is.null(colors)) colors <- viridis::plasma(10, end = 0.75)

    # R CMD Check variable bindings fix (see: https://bit.ly/3z24hbU) -----
    word <- freq <- NULL

    cli::cat_line()
    cli::cli_alert_info(paste0(
        "You need to expand your plot window for this function to work. ",
        "If the function produces warnings, try to maximize the plot window ",
        "and running it again."
    ))
    cli::cat_line()

    wordcloud::wordcloud(words = data$word, freq = data$freq,
                         min.freq = min_freq, max.words = max_words,
                         random.order = random_order, rot.per = rot_per,
                         colors = colors, ...)

    invisible(NULL)
}


R/word_frequency.R
#' Map terms from a corpus and compute its word frequency
#'
#' @description
#'
#' `r lifecycle::badge("experimental")`
#'
#' `word_frequency()` map terms from a vector using the [tm][tm::Corpus()]
#' package system and outputs a [tibble][dplyr::tibble()] summarizing word
#' frequencies.
#'
#' Please note that `word_frequency()` is required for other `iramuteqlike`
#' functions.
#'
#' @param x An [atomic vector][base::is.atomic()] (usually a
#'   [character][character()] object) with a corpus (a collection of documents
#'   containing text). Each element of `x` will represent a document (e.g., a
#'   response from a survey).
#' @param language (optional) a string indicating the corpus language. See
#'   [tm::SimpleCorpus()][tm::SimpleCorpus()] to learn more (default: `"en"`).
#' @param stopwords (optional) a string indicating the corpus language to be
#'   pass to [tm::stopwords()]. [tm::stopwords()] return various kinds of words
#'   related to a specific language to help clean a text analysis (default:
#'   `"en"`).
#' @param other_stopwords (optional) a string indicating other stop words to be
#' removed from the text analysis.
#'
#' @return A [`tibble`][tibble::tibble()] object with two columns:
#'
#' * `word`: with an unique set of words mapped from `x`.
#' * `freq`: with the absolute frequency of a word in `word`.
#'
#' @family Mapping functions
#' @export
#'
#' @examples
#' if (requireNamespace("friends", quietly = TRUE)) {
#'     word_frequency(head(friends::friends[[1]], 100))
#' }
#'
#' if (requireNamespace("stringi", quietly = TRUE)) {
#'     word_frequency(stringi::stri_rand_lipsum(5))
#' }
word_frequency <- function(x, language = "en", stopwords = language,
                           other_stopwords = NULL) {
    checkmate::assert_atomic(x, min.len = 1)
    checkmate::assert_string(language)
    checkmate::assert_string(stopwords, null.ok = TRUE)
    checkmate::assert_character(other_stopwords, min.chars = 1, min.len = 1,
                                null.ok = TRUE)

    # R CMD Check variable bindings fix (see: https://bit.ly/3z24hbU) -----
    . <- NULL

    # toSpace <- tm::content_transformer(
    #     function (x, pattern) gsub(pattern, " ", x))

    out <- data.frame(doc_id = seq_along(x), text = as.character(x)) %>%
        tm::DataframeSource() %>%
        tm::SimpleCorpus(control = list(language = language)) %>%
        tm::tm_map(tolower) %>%
        tm::tm_map(tm::removeNumbers) %>%
        tm::tm_map(tm::removePunctuation) %>%
        tm::tm_map(tm::stripWhitespace)

    if (!is.null(stopwords)) {
        out <- out %>%
            tm::tm_map(tm::removeWords, tm::stopwords(stopwords))

        if (!is.null(other_stopwords)) {
            out <- out %>%
                tm::tm_map(tm::removeWords, other_stopwords)
        }
    }

    out <- out %>%
        tm::TermDocumentMatrix() %>%
        as.matrix()

    sort(rowSums(out), decreasing = TRUE) %>%
        dplyr::tibble(word = names(.), freq = .)
}


R/export_to_iramuteq.R
#' Export a vector to an IRaMuTeQ formatted corpus file
#'
#' @description
#'
#' `r lifecycle::badge("experimental")`
#'
#' `export_to_iramuteq()` export atomic vectors to files with the format
#' required by IRaMuTeQ for text corpuses (a collection of documents containing
#' text).
#'
#' @param x An [atomic vector][base::is.atomic()] (usually a
#'   [character][character()] object) with a corpus (a collection of documents
#'   containing text). Each element of `x` will represent a document (e.g., a
#'   response from a survey).
#' @param dir (optional) a string indicating the directory on where the
#' corpus files must be saved (default: `utils::choose.dir()`).
#' @param file_name (optional) a string indicating the file name of the corpus.
#' @param var_name (optional) a string indicating the variable name for each
#' document. This is required by IRaMuTeQ for distinguish each document inside
#' a corpus (default: `"respondent"`).
#'
#' @return An invisible `NULL`. This function don't aim to return values.
#'
#' @family Utility functions
#' @export
#'
#' @examples
#' \dontrun{
#' if (requireNamespace("friends", quietly = TRUE)) {
#'     export_to_iramuteq(head(friends::friends[[1]], 100),
#'                        dir = tempdir(), file_name = "test.txt")
#' }
#' }
export_to_iramuteq <- function(x, dir = utils::choose.dir(),
                               file_name = paste0(deparse(substitute(x)),
                                                  ".txt"),
                               var_name = "respondent") {
    checkmate::assert_atomic(x, min.len = 1)
    checkmate::assert_directory_exists(dir)
    checkmate::assert_string(file_name, min.chars = 1)
    checkmate::assert_string(var_name, min.chars = 1,
                             pattern = "^[A-Za-z0-9.-_]+$")

    if (stringr::str_detect(file_name, "^[\\w\\-. ]+$", negate = TRUE)) {
        file_name <- "corpus.txt"

        cli::cli_alert_info(paste0(
            "The {.strong {cli::col_red('file_name')}} argument ",
            "have an invalid value. The file was named as ",
            "{.strong {cli::col_blue(file_name)}}."
        ))
    }

    file <- file.path(dir, file_name)

    readr::write_lines(iramuteq_format(x, var_name),
                       file, sep = "")

    readr::read_lines(file, skip_empty_rows = TRUE) %>%
        readr::write_lines(file)

    invisible(NULL)
}

iramuteq_format <- function(x, var_name = "respondent") {
    checkmate::assert_atomic(x, min.len = 1)
    checkmate::assert_string(var_name, min.chars = 1)

    # R CMD Check variable bindings fix (see: https://bit.ly/3z24hbU) -----
    . <- NULL

    x %>% as.character() %>%
        gutils:::na_replace() %>%
        tidy_case() %>%
        paste(paste0("**** *", var_name, "_", seq_along(.), "*"),
              ., sep = "\n", collapse = "\n")
}

tidy_case <- function(x) {
    checkmate::assert_character(x, min.len = 1)

    x %>% stringr::str_trim() %>%
        stringr::str_replace_all("\\n\\n|\\n\\n\\n|\\n\\n\\n\\n", "\n")
}


R/plot_word_frequency.R
#' Plot word frequencies
#'
#' @description
#'
#' `r lifecycle::badge("experimental")`
#'
#' `plot_word_frequency()` plots word frequencies charts related to an output of
#' [word_frequency()].
#'
#' @param data A [`tibble`][tibble::tibble()] object with two columns: `word`
#'   (with an unique set of words) and `freq` (with the absolute frequency of a
#'   word in `word`). Use [word_frequency()] to generate this value.
#' @param min_freq (optional) a number indicating the minimum frequency that a
#'   word can have to be plotted (default: `1`).
#' @param max_words (optional) a number indicating the maximum number of words
#'   to be plotted (default: `100`).
#' @param dist (optional) a [`logical`][base::logical()] value
#'   indicating if the function must plot a frequency _versus_ range chart
#'   for visualizing the whole distribution (default: `FALSE`).
#' @param color (optional) a string with a Hex color code indicating the color
#'   of the data (default: `viridis::viridis(1)`).
#'
#' @return An invisible `NULL`. This function don't aim to return values.
#'
#' @family Plot functions
#' @export
#'
#' @examples
#' if (requireNamespace("friends", quietly = TRUE)) {
#'     data <- word_frequency(head(friends::friends[[1]], 100))
#'     plot_word_frequency(data)
#'     suppressMessages(plot_word_frequency(data, dist = TRUE))
#'
#'     ## Note that graphics rendered in the documentation may
#'     ## not represent the quality of the output.
#' }
#'
#' if (requireNamespace("stringi", quietly = TRUE)) {
#'     data <- word_frequency(stringi::stri_rand_lipsum(5))
#'     plot_word_frequency(data)
#'     suppressMessages(plot_word_frequency(data, dist = TRUE))
#'
#'     ## Note that graphics rendered in the documentation may
#'     ## not represent the quality of the output.
#' }
plot_word_frequency <- function(data, min_freq = 1, max_words = 10,
                                dist = FALSE, color = viridis::viridis(1)) {
    checkmate::assert_data_frame(data, min.rows = 1, min.cols = 2)
    checkmate::assert_names(names(data), identical.to = c("word", "freq"))
    checkmate::assert_number(min_freq, lower = 1)
    checkmate::assert_number(max_words, lower = 1)
    checkmate::assert_flag(dist)
    checkmate::assert_string(color, pattern = "^#")

    # R CMD Check variable bindings fix (see: https://bit.ly/3z24hbU) -----
    word <- freq <- NULL

    if (isTRUE(dist)) {
        cli::cat_line()
        cli::cli_alert_info(paste0(
            "When {.strong {cli::col_blue('dist = TRUE')}} ",
            "{.strong {cli::col_red('min_freq')}} and ",
            "{.strong {cli::col_red('max_words')}} are ignored."
        ))
        cli::cat_line()

        plot <- data %>%
            dplyr::add_count(freq) %>%
            dplyr::mutate(range = seq(length(word))) %>%
            ggplot2::ggplot(ggplot2::aes(x = range, y = freq)) +
            ggplot2::geom_point(colour = color) +
            ggplot2::scale_x_log10() +
            ggplot2::scale_y_log10() +
            ggplot2::labs(x = "log(range)", y = "log(frequencies)")

        print(plot)
    } else {
        if (max_words > nrow(data)) max_words <- nrow(data)

        plot <- data %>%
            dplyr::filter(freq >= min_freq) %>%
            dplyr::slice(seq(1, max_words)) %>%
            ggplot2::ggplot(ggplot2::aes(x = freq,
                                         y = stats::reorder(word, freq))) +
            ggplot2::geom_col(fill = color) +
            ggplot2::labs(x = "Frequency", y = "Words")

        print(plot)
    }

    invisible(NULL)
}


R/plot.R
#' Plot a word cloud
#'
#' @description
#'
#' `r lifecycle::badge("experimental")`
#'
#' `plot_wordcloud()` plots a word cloud using an output of [word_frequency()].
#'
#' This function uses [wordcloud()][wordcloud::wordcloud()] to generate
#' word cloud graphics.
#'
#' @param data A [`tibble`][tibble::tibble()] object with two columns: `word`
#'   (with an unique set of words) and `freq` (with the absolute frequency of a
#'   word in `word`). Use [word_frequency()] to generate this value.
#' @param min_freq (optional) A number indicating the minimum frequency that a
#'   word can have to be plotted (default: `2`).
#' @param max_words (optional) A number indicating the maximum number of words
#'   to be plotted (default: `100`).
#' @param random_order (optional) A [`logical`][base::logical()] value
#'   indicating if the function must plot words in random order. If `FALSE`,
#'   the words will be plotted in decreasing frequency. See
#'   [wordcloud()][wordcloud::wordcloud()] to learn more (default: `FALSE`).
#' @param rot_per (optional) A number indicating the proportion words with 90
#'   degree rotation (default: `0.35`).
#' @param colors (optional) A [`character`][base::character()] object with Hex
#'   color codes indicating the color that the words must have, from least to
#'   most frequent. If `NULL`, the [viridis plasma palette][viridis::plasma()]
#'   will be used (default: `NULL`).
#' @param ... (optional) Additional parameters to be passed to the
#' [wordcloud()][wordcloud::wordcloud()] function.
#'
#' @return An invisible `NULL`. This function don't aim to return values.
#'
#' @family Plot functions
#' @export
#'
#' @examples
#' if(interactive()) {
#' if (requireNamespace("friends", quietly = TRUE)) {
#'     x <- word_frequency(head(friends::friends[[1]], 100))
#'     plot_wordcloud(x)
#' }
#'
#' if (requireNamespace("stringi", quietly = TRUE)) {
#'     x <- word_frequency(stringi::stri_rand_lipsum(5))
#'     plot_wordcloud(x)
#' }
#' }
plot_wordcloud <- function(data, min_freq = 2, max_words = 100,
                           random_order = FALSE, rot_per = 0.35,
                           colors = NULL, ...) {
    checkmate::assert_data_frame(data, min.rows = 1, min.cols = 2)
    checkmate::assert_names(names(data), identical.to = c("word", "freq"))
    checkmate::assert_integerish(min_freq, lower = 1)
    checkmate::assert_integerish(max_words, lower = 1)
    checkmate::assert_character(colors, min.len = 1, pattern = "^#",
                                null.ok = TRUE)

    if (!interactive()) {
        cli::cli_abort("This function can only be used in interactive mode.")
    }

    if (is.null(colors)) colors <- viridis::plasma(10, end = 0.75)

    # R CMD Check variable bindings fix (see: https://bit.ly/3z24hbU) -----
    word <- freq <- NULL

    cli::cat_line()
    cli::cli_alert_info(paste0(
        "You need to expand your plot window for this function to work. ",
        "If the function produces warnings, try to maximize the plot window ",
        "and running it again."
    ))
    cli::cat_line()

    wordcloud::wordcloud(words = data$word, freq = data$freq,
                         min.freq = min_freq, max.words = max_words,
                         random.order = random_order, rot.per = rot_per,
                         colors = colors, ...)

    invisible(NULL)
}

#' Plot word frequencies
#'
#' @description
#'
#' `r lifecycle::badge("experimental")`
#'
#' `plot_word_frequency()` plots word frequencies charts related to an output of
#' [word_frequency()].
#'
#' @param data A [`tibble`][tibble::tibble()] object with two columns: `word`
#'   (with an unique set of words) and `freq` (with the absolute frequency of a
#'   word in `word`). Use [word_frequency()] to generate this value.
#' @param min_freq (optional) A number indicating the minimum frequency that a
#'   word can have to be plotted (default: `1`).
#' @param max_words (optional) A number indicating the maximum number of words
#'   to be plotted (default: `100`).
#' @param dist (optional) A [`logical`][base::logical()] value
#'   indicating if the function must plot a frequency _versus_ range chart
#'   for visualizing the whole distribution (default: `FALSE`).
#' @param color (optional) A string with a Hex color code indicating the color
#'   of the data (default: `viridis::viridis(1)`).
#'
#' @return An invisible `NULL`. This function don't aim to return values.
#'
#' @family Plot functions
#' @export
#'
#' @examples
#' if(interactive()) {
#' if (requireNamespace("friends", quietly = TRUE)) {
#'     x <- word_frequency(head(friends::friends[[1]], 100))
#'     plot_word_frequency(x)
#'     plot_word_frequency(x, dist = TRUE)
#' }
#'
#' if (requireNamespace("stringi", quietly = TRUE)) {
#'     x <- word_frequency(stringi::stri_rand_lipsum(5))
#'     plot_word_frequency(x)
#'     plot_word_frequency(x, dist = TRUE)
#' }
#' }
plot_word_frequency <- function(data, min_freq = 1, max_words = 10,
                                dist = FALSE, color = viridis::viridis(1)) {
    checkmate::assert_data_frame(data, min.rows = 1, min.cols = 2)
    checkmate::assert_names(names(data), identical.to = c("word", "freq"))
    checkmate::assert_integerish(min_freq, lower = 1)
    checkmate::assert_integerish(max_words, lower = 1)
    checkmate::assert_flag(dist)
    checkmate::assert_string(color, min.len = 1, pattern = "^#")

    # R CMD Check variable bindings fix (see: https://bit.ly/3z24hbU) -----
    word <- freq <- NULL

    if (isTRUE(dist)) {
        cli::cat_line()
        cli::cli_alert_info(paste0(
            "When {.strong {cli::col_blue('dist = TRUE')}} ",
            "{.strong {cli::col_red('min_freq')}} and ",
            "{.strong {cli::col_red('max_words')}} are ignored."
        ))
        cli::cat_line()

        data %>%
            dplyr::add_count(freq) %>%
            dplyr::mutate(range = seq(length(word))) %>%
            ggplot2::ggplot(ggplot2::aes(x = range, y = freq)) +
            ggplot2::geom_point(colour = color) +
            ggplot2::scale_x_log10() +
            ggplot2::scale_y_log10() +
            ggplot2::labs(x = "log(range)", y = "log(frequencies)")
    } else {
        if (max_words > nrow(data)) max_words <- nrow(data)

        data %>%
            dplyr::filter(freq >= min_freq) %>%
            dplyr::slice(seq(1, max_words)) %>%
            ggplot2::ggplot(ggplot2::aes(x = freq,
                                         y = stats::reorder(word, freq))) +
            ggplot2::geom_col(fill = color) +
            ggplot2::labs(x = "Frequency", y = "Words")
    }
}
